### ISSUE TITLE
Can HTTP basic auth (with distinct credentials per 'user') work alongside Vouch?

### ISSUE THREAD

**Issue by mig5** [vouch/vouch-proxy#390]
Hi,

An organisation I'm implementing Vouch for, would like to *retain* HTTP basic auth over an application (with distinct credentials for separate users), *in addition* to using Vouch (rather than replacing the basic auth with Vouch on its own).

As far as I can tell, this is not possible, because of the design with Vouch: setting `error_page 401` to redirect to Vouch to perform oAuth, means that the 401 response sent from the server for Basic Auth (if none was provided in the request), creates a redirect loop by sending that 401 response to Vouch/the oAuth provider, over and over. That at least has been my experience when trying this out. The only way to get it to work was to disable Basic Auth end then Vouch worked perfectly (yay!).

I just wanted to get confirmation that essentially, Vouch is incompatible with basic HTTP auth (such as `auth_basic` in Nginx) at least when implemented *in conjunction*, because it needs to 'hijack' the 401 response. Am I right? 

I know there are other issues such as https://github.com/vouch/vouch-proxy/issues/360, but they seem to be different: they are hardcoding some static dummy user/pass (or at least the password) for all users, whereas my organisation wants to keep the distinct auth as a sort of 'failsafe' in case a Vouch implementation issue leaves the app accidentally exposed.

Thanks!

**Comment by bnfinet** [comment]
I haven't done it personally but I don't see a reason why you shouldn't be able to use `auth_basic` and `auth_request` together with `satisfy all;`

Some of these more advanced examples of using Basic Auth with nginx may be of use to you.  Particularly the `satisfy all;` or `satisfy any;` directives.

https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-http-basic-authentication/#combine

If you do develop a solution based on that nginx setup/syntax could you please document the bones of it hear for the common good?

**Comment by mig5** [comment]
Thanks! But is it not true that by having `error_page 401`, the 401 HTTP response for basic auth (to prompt the user to enter basic auth credentials) will get 'intercepted' by the 302 redirect to Vouch, in the standard Vouch configuration? I can't see how using `satisfy` will work around that - except for the scenario whereby the user has somehow already sent HTTP authentication header in their request (without being prompted first).

**Comment by bnfinet** [comment]
Really I'm not sure.  I haven't had a need for that setup so I haven't explored that.

If you put your Basic Auth first does that work?  Is there a reason that Vouch Proxy needs to be first?

**Comment by mig5** [comment]
Thanks @bnfinet for your quick responses.

I have `auth_request /validate;` in the 'main' part of my `server` Nginx block (along with the `error_page 401` etc all as per standard Vouch config).

I have auth_basic set in a `location /` block:

```
  location / {
       auth_basic "restricted";
       auth_basic_user_file auth_basic/htpasswd_file;
       [....]
   }
```

It doesn't seem to make a difference if I move `auth_basic` out of the `location` block and to be 'first' above the `auth_request` line.

Likewise, if I move `auth_request` into `location /` and `auth_basic` into the upper server context, it's the same problem. Besides, I want to make sure `auth_request` works on other locations too, not just the `location /`

Are you suggesting to try and put `auth_request` on some more specific route than `/` ? Which of course would depend on if the app has some other 'entry' point ?

Thanks again for your helpful responses.

**Comment by bnfinet** [comment]
I'm not sure I'm following your configuration.  Could you please provide your configs in the manner described in the README?

Have you consulted the nginx documentation for `satisfy all;`?  I feel like you should just try a bunch of stuff and see what works.  I'm not sure I'm the best resource for you to mine for success.

**Comment by mig5** [comment]
@bnfinet thanks again - I have put my vhost here https://gist.github.com/mig5/238660e43f3af568adabe8ce329011bb . You'll see it's really very simple and standard. The only unique thing is the `auth_basic` stuff in the location block.

As mentioned above (I had edited my post), `satisfy all` doesn't help. I really suspect that `error_page 401` intercepts the HTTP auth popup box and redirects to Vouch and the loop begins.

I definitely didn't intend to 'mine' you for support. As an opensource developer myself I understand the privilege you're already providing me with responses up til now. Just thought there might be a trick and surprised others haven't run into this. Thanks again. If I find a way to make it work I will be sure to provide docs.

**Comment by bnfinet** [comment]
how about stacked instead of flat?

If you put a basic_auth proxy in front of VP would that get you there?

**Comment by bnfinet** [comment]
@mig5 any progress?  What if you put a basic auth reverse proxy in front of your vouch protected app (also rev proxy)?

**Comment by mig5** [comment]
@bnfinet sorry for the lack of reply, managed to type a response and failed to submit it..

That would indeed work for having HTTP auth *somewhere* in the layout. We considered that after your suggestion, but in the end, we decided that what we were trying to mitigate was any risk that the actual `auth_request` call that redirects to Vouch in the first place, might fail for some unknown reason.

In the end, the app that we were trying to protect on the backend, has its own native authentication scheme (Elasticsearch/Kibana, in our case). So we decided to just rely on *that* and the standard Vouch setup, and drop basic HTTP auth altogether. Fortunately, Kibana's native auth system doesn't throw a 401 if not logged in - it does its own 302 to a login screen.

So our setup is now a bit more traditional:

1) User visits Kibana but is not auth'd to it nor Vouch
2) User gets redirected to Vouch/the oAuth IDP
3) User auths at the oAuth IDP and is redirected back to Kibana
4) User is still not logged in at Kibana so is 302'd to the Kibana authentication screen
5) User signs in to Kibana with native auth creds.

Not quite the original vision we had, but solves our needs for now. If I were to do it again elsewhere, I'd do it like you say, and put HTTP auth over the Vouch service, but that still carries the implicit assumption that the redirect to Vouch itself has worked, e.g there hasn't been any sort of regression in the vhost, or bug in Nginx's `auth_request` functionality, that has broken it somehow.

Thanks once again for all your help and suggestions - it really helped us. Loving rolling out Vouch despite these hurdles. I'll close this out!
